HealthVault/Classes/Caching/MHVThingCache.m

```
@@ -1,20 +1,471 @@
+        }
+    }
+    return self;
+}
+
+- (void)startCache
+{
+    self.syncTimer = [NSTimer scheduledTimerWithTimeInterval:_cacheConfiguration.syncIntervalSeconds
+                                                      target:self
+                                                    selector:@selector(syncTimerAction)
+                                                    userInfo:nil
+                                                     repeats:YES];
+    
+    [self prepareCacheForRecords:self.connection.personInfo.records completion:^(NSError * _Nullable error)
+     {
+         if (error)
+         {
+             MHVLOG(@"ThingCache: Error preparing cache: %@", error.localizedDescription);
+         }
+         
+         //Start a sync so cache will get updated when app launched & user authenticated
+         [self syncWithCompletionHandler:^(NSInteger syncedItemCount, NSError *error)
+          {
+              if (error)
+              {
+                  MHVLOG(@"ThingCache: Error performing sync: %@", error.localizedDescription);
+              }
+          }];
+     }];
+}
+
+- (NSError *_Nullable)clearAllCachedData
+{
+    NSError *error = [self.database deleteDatabase];
+    if (error)
+    {
+        MHVLOG(@"ThingCache: Error deleting database: %@", error.localizedDescription);
+    }
+
+    @synchronized (self.syncCompletionHandlers)
+    {
+        [self.syncCompletionHandlers removeAllObjects];
+    }
+    
+    if (self.syncTimer.isValid)
+    {
+        [self.syncTimer invalidate];
+    }
+    
+    return error;
+}
+
+#pragma mark - Cached Results
+
+- (void)cachedResultsForQueries:(MHVThingQueryCollection *)queries
+                       recordId:(NSUUID *)recordId
+                     completion:(void(^)(MHVThingQueryResultCollection *_Nullable resultCollection, NSError *_Nullable error))completion;
+{
+    id<MHVCachedRecord> record = [self.database fetchCachedRecord:recordId.UUIDString];
```

We need to add a check here to see if the database is ready to use. It should probably return an error that the thing client can use to issue the request to HealthVault until the database has been synchronized.